eval/topdown.go

```diff
@@ -173,6 +173,51 @@ func dereferenceVar(v opalog.Var, ctx *TopDownContext) (interface{}, error) {
 	return valueToInterface(binding, ctx)
 }
 
+func evalContext(ctx *TopDownContext, iter TopDownIterator) error {
+
+	if ctx.Index >= len(ctx.Query) {
+		return iter(ctx)
+	}
+
+	if ctx.Current().Negated {
+		return evalContextNegated(ctx, iter)
+	}
+
+	return evalTerms(ctx, func(ctx *TopDownContext) error {
+		return evalExpr(ctx, func(ctx *TopDownContext) error {
+			ctx = ctx.Step()
+			return evalContext(ctx, iter)
+		})
+	})
+}
+
+func evalContextNegated(ctx *TopDownContext, iter TopDownIterator) error {
+
+	negation := &TopDownContext{
+		Query:    opalog.Body([]*opalog.Expr{ctx.Current().Complement()}),
+		Bindings: ctx.Bindings,
+		Previous: ctx,
+		Store:    ctx.Store,
+	}
+
+	isDefined := false
+
+	err := evalContext(negation, func(*TopDownContext) error {
+		isDefined = true
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	if !isDefined {
+		return evalContext(ctx.Step(), iter)
+	}
+
+	return nil
```

Comment:
What does the return value of this function represent?  I tried looking at higher-level eval routines, but couldn't seem to track down anything that used the return value.